README.md
[demo]: https://raw.githubusercontent.com/jabney/vector-keyframes/master/meta/demo.gif

# vector-keyframes
Interpolate between any number of scalar, 2d, or 3d vector keyframes with a focus on colors

[![Build Status](https://travis-ci.org/jabney/vector-keyframes.svg?branch=master)](https://travis-ci.org/jabney/vector-keyframes)[![codecov](https://codecov.io/gh/jabney/vector-keyframes/branch/master/graph/badge.svg)](https://codecov.io/gh/jabney/vector-keyframes)

## Installation

    npm install vector-keyframes

or

    yarn add vector-keyframes

![keyframe demo][demo]

*[demo app](https://jabney.github.io/vector-keyframes-demo/dist/index.html)*

## Examples
See [examples.js](https://github.com/jabney/vector-keyframes/blob/master/src/example.js)

```javascript
// Import the module and create some library aliases.
const vk = require('vector-keyframes')
const vector3d = vk.vector3d
const util = vk.util

// Define some keyframes ordered by stop, low to high.
const keyframes = [{
  stop: 0,
  value: [0, 0, 255]
}, {
  stop: 0.5,
  value: [0, 255, 0]
}, {
  stop: 1,
  value: [0, 0, 255]
}]

// Pass the keyframes and the time parameter to interpolate.
const result = vector3d.keyframeInterpolate(keyframes, 0.25)
console.log('vector    (t=0.25):', result)
console.log('rgb color (t=0.25):', util.vector3dToRgb(result))
console.log('hex color (t=0.25):', util.vector3dToHex(result))
```

### scalar
`scalar` provides a set of interpolation methods that `vector2d` and `vector3d` are built on.

```javascript
const vk = require('vector-keyframes')
const scalar = vk.scalar

const keyframes = [{
  stop: 0,
  value: 0
}, {
  stop: 0.5,
  value: 1
}, {
  stop: 1,
  value: 0
}]

console.log('linear   (t=0):',
  scalar.keyframeInterpolate(keyframes, 0))

console.log('linear (t=0.5):',
  scalar.keyframeInterpolate(keyframes, 0.5))

console.log('linear   (t=1):',
  scalar.keyframeInterpolate(keyframes, 1))

console.log('linear (t=0.1):',
  scalar.keyframeInterpolate(keyframes, 0.1))

console.log('smooth (t=0.1):',
  scalar.keyframeInterpolate(keyframes, 0.1, 'smooth'))
```

Output:

```
linear   (t=0): 0
linear (t=0.5): 1
linear   (t=1): 0
linear (t=0.1): 0.2
smooth (t=0.1): 0.10400000000000002
```

### vector2d
`vector2d` iterpolates between 2d vectors.

```javascript
const vk = require('vector-keyframes')
const vector2d = vk.vector2d

const keyframes = [{
  stop: 0,
  value: [0, 100]
}, {
  stop: 0.5,
  value: [100, 100]
}, {
  stop: 1,
  value: [100, 0]
}]

console.log('linear   (t=0):',
  vector2d.keyframeInterpolate(keyframes, 0))

console.log('linear (t=0.5):',
  vector2d.keyframeInterpolate(keyframes, 0.5))

console.log('linear   (t=1):',
  vector2d.keyframeInterpolate(keyframes, 1))

console.log('linear (t=0.1):',
  vector2d.keyframeInterpolate(keyframes, 0.1))

console.log('smooth (t=0.1):',
  vector2d.keyframeInterpolate(keyframes, 0.1, 'smooth'))
```

Output:

```
linear   (t=0): [ 0, 100 ]
linear (t=0.5): [ 100, 100 ]
linear   (t=1): [ 100, 0 ]
linear (t=0.1): [ 20, 100 ]
smooth (t=0.1): [ 10.400000000000002, 100 ]
```

### vector3d
`vector3d` interpolates between 3d vectors. The `util` library provides some conversion helpers.

```javascript
const vk = require('vector-keyframes')
const vector3d = vk.vector3d
const util = vk.util

const keyframes = [{
  stop: 0,
  value: [0, 0, 255]
}, {
  stop: 0.5,
  value: [0, 255, 0]
}, {
  stop: 1,
  value: [0, 0, 255]
}]

let result = vector3d.keyframeInterpolate(keyframes, 0.25)
console.log('\nvector    (t=0.25):', result)
console.log('rgb color (t=0.25):', util.vector3dToRgb(result))
console.log('hex color (t=0.25):', util.vector3dToHex(result))

result = vector3d.keyframeInterpolate(keyframes, 0.5)
console.log('\nvector    (t=0.5):', result)
console.log('rgb color (t=0.5):', util.vector3dToRgb(result))
console.log('hex color (t=0.5):', util.vector3dToHex(result))

result = vector3d.keyframeInterpolate(keyframes, 0.9)
console.log('\nvector    (t=0.9):', result)
console.log('rgb color (t=0.9):', util.vector3dToRgb(result))
console.log('hex color (t=0.9):', util.vector3dToHex(result))
```

Output:

```
vector    (t=0.25): [ 0, 127.5, 127.5 ]
rgb color (t=0.25): rgb(0,128,128)
hex color (t=0.25): #008080

vector    (t=0.5): [ 0, 255, 0 ]
rgb color (t=0.5): rgb(0,255,0)
hex color (t=0.5): #00ff00

vector    (t=0.9): [ 0, 51, 204 ]
rgb color (t=0.9): rgb(0,51,204)
hex color (t=0.9): #0033cc
```

### custom
Custom interpolation can be achieved by providing a base library and defining
certain methods within:

```javascript
const vk = require('vector-keyframes')
const scalar = vk.scalar

// Create a base 'alpha' library which defines its own interpolation
// methods, making use of the ones in the 'scalar' lib.
const alpha = {

  zero() {
    return '*'
  },

  linearInterpolate(a, b, t) {
    return vk.scalar.linearInterpolate(a.charCodeAt(0), b.charCodeAt(0), t)
  },

  smoothInterpolate(a, b, t) {
    return vk.scalar.smoothInterpolate(a.charCodeAt(0), b.charCodeAt(0), t)
  },

  keyframeInterpolate(keyframes, t, timing='linear', lib=alpha) {
    const result = vk.scalar.keyframeInterpolate(keyframes, t, timing, lib)
    return String.fromCharCode(Math.round(result))
  }
}

// Use the newly defined 'alpha' library.
const keyframes = [{
  stop: 0,
  value: 'a'
}, {
  stop: 0.5,
  value: 'z'
}, {
  stop: 1,
  value: 'a'
}]

console.log('linear   (t=0):',
  alpha.keyframeInterpolate(keyframes, 0, 'linear'))

console.log('linear (t=0.5):',
  alpha.keyframeInterpolate(keyframes, 0.5, 'linear'))

console.log('linear   (t=1):',
  alpha.keyframeInterpolate(keyframes, 1, 'linear'))

console.log('linear (t=0.1):',
  alpha.keyframeInterpolate(keyframes, 0.1, 'linear'))

console.log('smooth (t=0.1):',
  alpha.keyframeInterpolate(keyframes, 0.1, 'smooth'))
```

Output:

```
linear   (t=0): a
linear (t=0.5): z
linear   (t=1): a
linear (t=0.1): f
smooth (t=0.1): d
```


CHANGELOG.md
# Change log for Vector Keyframes

## [v0.2.16] - 2019-06-07
npm audit fix.

## [v0.2.15] - 2019-05-03
Remove `node 4` from `travis.yml`.

## [v0.2.14] - 2019-05-03
npm audit fix

## [v0.2.13] - 2019-05-03
npm audit fix

## [v0.2.12] - 2017-07-18
Dependency reconfiguration

## [v0.2.11] - 2017-07-18
Add codecov badge

## [v0.2.10] - 2017-07-18
Add code coverage reports to test script

## [v0.2.9] - 2017-07-18
Update readme

## [v0.2.8] - 2017-07-18
Update readme

## [v0.2.7] - 08-05-2017
### Changed
- Clamp time value in smooth method to [0, 1]
- Add a test spec for custom interplators

## [v0.2.6] - 06-05-2017
### Changed
- Don't clamp time parameter in keyframeInterpolate
- JSDoc comments

## [v0.2.5] - 06-05-2017
### Changed
- Remove binary search

## [v0.2.4] - 05-05-2017
### Changed
- Use alternate value midpoint calculation in tweenSearch

## [v0.2.3] - 05-05-2017
### Changed
- Fixed issue where custom interpolator falls down when tween search
  returns a single keyframe

## [v0.2.2] - 04-05-2017
### Changed
- Implement tweenSearch for keyframeInterpolate

## [v0.1.4] - 01-05-2017
### Changed
- Minor refactoring
- Update readme and example.js

## [v0.1.3] - 01-05-2017
### Changed
- Prebuild project

## [v0.1.2] - 01-05-2017
### Changed
- Update .npmignore

## [v0.1.1] - 01-05-2017
### Changed
- Update dependencies in package.json

## [v0.1.0] - 01-05-2017
### Initial Release
- src/index.js: one-and-only source file
- src/test/*_spec.js: unit tests
- example.js: usage examples


test/util.js
var assert = require('assert')
var util = require('../index').util

describe('utilities', function () {

  describe('linearInterpolate', function () {

    it('interpolates between two scalars', function () {
      var start = 1
      var end = 2
      var result

      result = util.linearInterpolate(start, end, 0)
      assert.equal(result, start, 't=0 results in start value')

      result = util.linearInterpolate(start, end, 1)
      assert.equal(result, end, 't=1 results in end value')

      result = util.linearInterpolate(start, end, 0.5)
      assert.equal(result, (start+end)/2, 't=0.5 results in start+end midpoint')
    })
  })

  describe('smoothInterpolate', function () {
    it('interpolates between two scalars', function () {
      var start = 1
      var end = 2
      var result

      result = util.linearInterpolate(start, end, 0)
      assert.equal(result, start, 't=0 results in start value')

      result = util.linearInterpolate(start, end, 1)
      assert.equal(result, end, 't=1 results in end value')

      result = util.linearInterpolate(start, end, 0.5)
      assert.equal(result, (start+end)/2, 't=0.5 results in start+end midpoint')
    })

    it('has slow start and fast finish', function () {
      var start = 1
      var end = 2
      var smooth, linear

      smooth = util.smoothInterpolate(start, end, 0.1)
      linear = util.linearInterpolate(start, end, 0.1)
      assert(smooth < linear, 'starts slower than linear')

      smooth = util.smoothInterpolate(start, end, 0.9)
      linear = util.linearInterpolate(start, end, 0.9)
      assert(smooth > linear, 'finishes faster than linear')
    })
  })

  describe('vectorToHex', function () {
    it('converts 3d byte vector to 6-digit hex color', function () {
      var vector, hex

      vector = [0, 0, 0]
      hex = util.vectorToHex(vector)
      assert.equal(hex, '#000000')

      vector = [127, 127, 127]
      hex = util.vectorToHex(vector)
      assert.equal(hex, '#7f7f7f')

      vector = [255, 255, 255]
      hex = util.vectorToHex(vector)
      assert.equal(hex, '#ffffff')
    })
  })

  describe('hexToVector', function () {
    it('converts 6-digit hex color to a 3d byte vector', function () {
      var hex, vector

      hex = '#000000'
      vector = util.hexToVector(hex)
      assert.deepEqual(vector, [0, 0, 0])

      hex = '#7f7f7f'
      vector = util.hexToVector(hex)
      assert.deepEqual(vector, [127, 127, 127])

      hex = '#ffffff'
      vector = util.hexToVector(hex)
      assert.deepEqual(vector, [255, 255, 255])
    })
  })

  describe('vectorToRgb', function () {
    it('converts vector to rgb color string', function () {
      var rgbColor = util.vectorToRgb([0,127.4,254.6])
      assert.equal(rgbColor, 'rgb(0,127,255)')
    })
  })

  describe('rgbToVector', function () {
    it('converts rgb color string to vector', function () {
      var vector = util.rgbToVector('rgb( 0,127, 255)')
      assert.deepStrictEqual(vector, [0,127,255])
    })
  })
})


test/search.js
var assert = require('assert')
var search = require('../index').search
var _ = require('lodash')

describe('binary search', function () {
  it('returns null for an empty list', function () {
    var list = []
    var found = search.binary(list, 1)
    assert.equal(found, null)
  })

  it('returns null for target not in list', function () {
    var list = [1, 2, 3, 4, 5, 6, 7]
    var found = search.binary(list, 8)
    assert.equal(found, null)
  })

  it('finds all targets in lists of length [1,100]', function () {
    var maxSize = 1e2
    var skip = 1
    var list, size, find, found

    // Generate lists of size [1,maxSize]
    for (size = 1; size <= maxSize; size += skip) {
      list = _.range(0, size, skip)

      for (find = 0; find < list.length; find += skip) {
        found = search.binary(list, find)
        assert.equal(find, found)
      }
    }
  })

  it('finds values in O(lg(n)) time', function () {
    var maxSize = 1e3
    var skip = 100
    var list, size, find, found, compares, maxCompares

    // Generate lists of size [skip,maxSize]
    for (size = skip; size <= maxSize; size += skip) {
      list = _.range(0, size)
      maxCompares = 0

      for (find = 0; find < size; ++find) {
        compares = 0

        found = search.binary(list, find, function (candidate, target) {
          maxCompares = Math.max(maxCompares, ++compares)

          if (target < candidate) {
            return -1
          } else if (target > candidate) {
            return 1
          } else {
            return 0
          }
        })

        assert.equal(find, found, 'found the target value')

        assert(maxCompares <= Math.floor(Math.log2(size)) + 1,
          'maxCompares <= log2(list_size) + 1')
      }
    }
  })
})



test/vector2d.js
var assert = require('assert')
var vector2d = require('../index').vector2d

describe('vector2d', function () {

  describe('zero', function () {

    it('returns a 2d zero vector', function () {
      assert.deepEqual(vector2d.zero(), [0, 0])
    })
  })

  describe('linearInterpolate', function () {
    it('interpolates between two 2d vectors', function () {
      var start = [0, 1]
      var end = [1, 2]
      var result

      result = vector2d.linearInterpolate(start, end, 0)
      assert.deepEqual(result, start, 't=0 results in start vector')

      result = vector2d.linearInterpolate(start, end, 1)
      assert.deepEqual(result, end, 't=1 results in end vector')

      result = vector2d.linearInterpolate(start, end, 0.5)
      assert.deepEqual(result, [0.5, 1.5],
        't=0.5 results in start+end midpoint')
    })
  })

  describe('smoothInterpolate', function () {
    it('interpolates between two 2d vectors', function () {
      var start = [0, 1]
      var end = [1, 2]
      var result

      result = vector2d.linearInterpolate(start, end, 0)
      assert.deepEqual(result, start, 't=0 results in start vector')

      result = vector2d.linearInterpolate(start, end, 1)
      assert.deepEqual(result, end, 't=1 results in end vector')

      result = vector2d.linearInterpolate(start, end, 0.5)
      assert.deepEqual(result, [0.5, 1.5],
        't=0.5 results in start+end midpoint')
    })

    it('has slow start and fast finish', function () {
      var start = [0, 1]
      var end = [1, 2]
      var smooth, linear

      smooth = vector2d.smoothInterpolate(start, end, 0.1)
      linear = vector2d.linearInterpolate(start, end, 0.1)
      assert(smooth[0] < linear[0], 'starts slower than linear')
      assert(smooth[1] < linear[1], 'starts slower than linear')

      smooth = vector2d.smoothInterpolate(start, end, 0.9)
      linear = vector2d.linearInterpolate(start, end, 0.9)
      assert(smooth[0] > linear[0], 'finishes faster than linear')
      assert(smooth[1] > linear[1], 'finishes faster than linear')
    })
  })

  describe('keyframeInterpolate', function () {

    it('returns the zero vector for empty keyframes', function () {
      var vector

      vector = vector2d.keyframeInterpolate([], 0)
      assert.deepEqual(vector, [0,0])

      vector = vector2d.keyframeInterpolate([], 0.5)
      assert.deepEqual(vector, [0,0])

      vector = vector2d.keyframeInterpolate([], 1)
      assert.deepEqual(vector, [0,0])
    })

    it('returns the first keyframe when only one is present', function () {
      var keyframes = [{
        stop: 0,
        vector: [255,255]
      }]
      var vector

      vector = vector2d.keyframeInterpolate(keyframes, 0)
      assert.deepEqual(vector, [255,255])

      vector = vector2d.keyframeInterpolate(keyframes, 0.5)
      assert.deepEqual(vector, [255,255])

      vector = vector2d.keyframeInterpolate(keyframes, 1)
      assert.deepEqual(vector, [255,255])
    })

    it('interpolates between two keyframes', function () {
      var keyframes = [{
        stop: 0,
        vector: [0,0]
      }, {
        stop: 1,
        vector: [255,255]
      }]
      var vector

      vector = vector2d.keyframeInterpolate(keyframes, 0)
      assert.deepEqual(vector, [0,0])

      vector = vector2d.keyframeInterpolate(keyframes, 0.5)
      assert.deepEqual(vector, [127.5,127.5])

      vector = vector2d.keyframeInterpolate(keyframes, 1)
      assert.deepEqual(vector, [255,255])
    })

    it('interpolates between three keyframes', function () {
      var keyframes = [{
        stop: 0,
        vector: [0,0]
      }, {
        stop: 0.5,
        vector: [48,96]
      }, {
        stop: 1,
        vector: [255,255]
      }]
      var vector

      vector = vector2d.keyframeInterpolate(keyframes, 0)
      assert.deepEqual(vector, [0,0])

      vector = vector2d.keyframeInterpolate(keyframes, 0.5)
      assert.deepEqual(vector, [48,96])

      vector = vector2d.keyframeInterpolate(keyframes, 1)
      assert.deepEqual(vector, [255,255])
    })

    it('clamps first and last keyframes', function () {
      var keyframes = [{
        stop: 0.25,
        vector: [0,0]
      }, {
        stop: 0.5,
        vector: [127,127]
      }, {
        stop: 0.75,
        vector: [255,255]
      }]
      var vector

      vector = vector2d.keyframeInterpolate(keyframes, 0)
      assert.deepEqual(vector, [0,0])

      vector = vector2d.keyframeInterpolate(keyframes, 0.25)
      assert.deepEqual(vector, [0,0])

      vector = vector2d.keyframeInterpolate(keyframes, 0.5)
      assert.deepEqual(vector, [127,127])

      vector = vector2d.keyframeInterpolate(keyframes, 0.75)
      assert.deepEqual(vector, [255,255])

      vector = vector2d.keyframeInterpolate(keyframes, 1)
      assert.deepEqual(vector, [255,255])
    })

    it('interpolates smooth or linear', function () {
      var keyframes = [{
        stop: 0,
        vector: [0,0]
      }, {
        stop: 1,
        vector: [255,255]
      }]
      var smooth, linear

      smooth = vector2d.keyframeInterpolate(keyframes, 0, 'smooth')
      linear = vector2d.keyframeInterpolate(keyframes, 0, 'linear')
      assert.deepEqual(smooth, [0,0])
      assert.deepEqual(linear, [0,0])

      smooth = vector2d.keyframeInterpolate(keyframes, 0.1, 'smooth')
      linear = vector2d.keyframeInterpolate(keyframes, 0.1, 'linear')
      assert(smooth[0] < linear[0])
      assert(smooth[1] < linear[1])

      smooth = vector2d.keyframeInterpolate(keyframes, 0.5, 'smooth')
      linear = vector2d.keyframeInterpolate(keyframes, 0.5, 'linear')
      assert.deepEqual(smooth, [127.5,127.5])
      assert.deepEqual(linear, [127.5,127.5])

      smooth = vector2d.keyframeInterpolate(keyframes, 0.9, 'smooth')
      linear = vector2d.keyframeInterpolate(keyframes, 0.9, 'linear')
      assert(smooth[0] > linear[0])
      assert(smooth[1] > linear[1])

      smooth = vector2d.keyframeInterpolate(keyframes, 1, 'smooth')
      linear = vector2d.keyframeInterpolate(keyframes, 1, 'linear')
      assert.deepEqual(smooth, [255,255])
      assert.deepEqual(linear, [255,255])
    })
  })
})


src/index.js

// -----------------------------------
// Tween search
// -----------------------------------

function tweenComparator(a, b, time) {
  if (a.stop <= time && time < b.stop) {
    return 0
  } else {
    if (time < a.stop) {
      return -1
    } else {
      return 1
    }
  }
}

function tweenSearch(sortedList, time, low, high, comparator) {
  let mid = Math.floor(low + ((high-low) / 2))
  let len = sortedList.length

  if (len == 0) return null
  if (mid < 0) return [sortedList[0]]
  if (mid == len-1) return [sortedList[len-1]]

  let a = sortedList[mid]
  let b = sortedList[mid+1]

  let result = comparator(a, b, time)

  if (result == 0) {
    return [a, b]
  } else {
    if (result < 0) {
      return tweenSearch(sortedList, time, low, mid-1, comparator)
    } else {
      return tweenSearch(sortedList, time, mid+1, high, comparator)
    }
  }
}

// -----------------------------------
// Keyframe interpolation
// -----------------------------------

function timingToInterpFn(timing, lib) {
  let map = {
    'linear': lib.linearInterpolate,
    'smooth': lib.smoothInterpolate
  }
  return map[timing] || map.linear
}

function keyframeInterpolate(keyframes, time, timing, lib) {

  if (!Array.isArray(keyframes) || !keyframes.length) {
    return lib.zero()
  }

  const result = search.tween(keyframes, time)
  const interpolate = timingToInterpFn(timing, lib)

  if (result.length == 1) {
    let value = result[0].value
    return interpolate(value, value, 0)
  }

  const [a, b] = result
  const t_segment = (time - a.stop) / (b.stop - a.stop)
  return interpolate(a.value, b.value, t_segment)
}

// -------------------------------------------------
// search library
// -------------------------------------------------

const search = {

  /**
   * Find the keyframe or keyframe pair closest to a given time
   * @param {keyframe[]} sortedList a list of keyframes sorted
   * by 'stop', e.g, [{stop: 0.1}, {stop: 0.3}, {stop: 0.5}]
   * @param {number} time the time nearest to the desired keyframe
   * or keyframe pair
   * @param {function} comparator the search comparator function, e.g.,
   * (candidate:keyframe, target:keyframe) => number
   */
  tween(sortedList, time, comparator=tweenComparator) {
    return tweenSearch(sortedList, time, 0, sortedList.length - 1, comparator)
  }
}

// -------------------------------------------------
// util library
// -------------------------------------------------

const util = {

  /**
   * Smooth a parameter in the range 0 <= t <= 1 in a manner
   * that approximates cosine.
   * @param {float} t interpolation parameter in the range [0, 1]
   * @returns {float}
   */
  smooth(t) {
    const time = Math.max(t, Math.min(t, 1), 0)
    return 3 * Math.pow(time, 2) - 2 * Math.pow(time, 3)
  },

  /**
   * Convert a 3d byte vector to a hex color string,
   * e.g, [0, 127, 255] -> #007fff
   * @param {number[]} vector a three-element array
   * @returns {string}
   */
  vector3dToHex(vector) {
    const hexColor = '#' + vector.slice(0,3).map((c) => {
      const byte = Math.round(c)
      const hex = byte.toString(16)
      return hex.length == 1 ? '0' + hex : hex
    }).join('')

    return hexColor
  },

  /**
   * Convert a hex color string to a 3d byte vector,
   * e.g, #007fff -> [0, 127, 255]
   * @param {string} hexColor
   * @returns {number[]}
   */
  hexToVector3d(hexColor) {
    const hex = hexColor.slice(1)
    const r = parseInt(hex.slice(0, 2), 16)
    const g = parseInt(hex.slice(2, 4), 16)
    const b = parseInt(hex.slice(4, 6), 16)
    return [r, g, b]
  },

  /**
   * Convert a 3d byte vector to an rgb color string,
   * e.g., [0, 127, 255] -> 'rgb(0,127,255)'
   * @param {number[]}
   * @returns {string}
   */
  vector3dToRgb(vector) {
    vector = vector.slice(0,3).map(function (byte) {
      return Math.round(byte)
    })
    return 'rgb(' + vector + ')'
  },

  /**
   * Convert an rgb color string to a 3d byte vector,
   * e.g., 'rgb(0,127,255)' -> [0, 127, 255]
   * @param {string} rgbColor
   * @returns {number[]}
   */
  rgbToVector3d(rgbColor) {
    var list = rgbColor.split(/[(), ]+/)
    return list.slice(1,4).map(function (s) {
      return +s
    })
  },

  /**
   * Sort a list of keyframes in place, or any list of objects
   * with a numerical 'stop' property
   * @param {keyframe[]} keyframes a list of keyframes in
   * the form, e.g.,  [{stop: 0.5}, {stop: 0.1}, {stop: 0.95}]
   * @returns {keyframe[]}
   */
  keyframeSort(keyframes) {
    return keyframes.sort((a, b) => a.stop - b.stop)
  }
}

// -------------------------------------------------
// scalar library
// -------------------------------------------------

const scalar = {

  /**
   * Return zero
   * @returns {number}
   */
  zero() {
    return 0
  },

  /**
   * Interpolate between a and b for a factor of t
   * @param {number} a the start value
   * @param {number} b the end value
   * @param {number} t the time parameter
   * @returns {number} the interpolated value
   */
  linearInterpolate(a, b, t) {
    return a + t*(b-a)
  },

  /**
   * Interpolate between a and b for a factor of t
   * with a smooth, ease in, ease out type effect
   * @param {number} a the start value
   * @param {number} b the end value
   * @param {number} t the time parameter
   * @returns {number} the interpolated value
   */
  smoothInterpolate(a, b, t) {
    return a + util.smooth(t)*(b-a)
  },

  /**
   * Interpolate with keyframes, e.g.,
   * [{stop: 0.1, value: 10}, {stop: 0.5, value: 20}, {stop: 0.9, value: 30}]
   * @param {keyframe[]} keyframes a list of keyframes
   * @param {number} time the time nearest the keyframe stop
   * @param {'linear'|'smooth'} timing interpolation timing
   * @param {object} lib the interpolation library to use
   */
  keyframeInterpolate(keyframes, time, timing='linear', lib=scalar) {
    return keyframeInterpolate(keyframes, time, timing, lib)
  }
}

// -------------------------------------------------
// vector2d library
// -------------------------------------------------

const vector2d = {

  /**
   * Return the zero vector
   */
  zero() {
    return [0, 0]
  },

  /**
   * Interpolate between vectors a and b for a factor of t
   * @param {number} a the start value
   * @param {number} b the end value
   * @param {number} t the time parameter
   * @returns {number} the interpolated value
   */
  linearInterpolate(a, b, t) {
    const lerp = scalar.linearInterpolate
    return [
      lerp(a[0], b[0], t),
      lerp(a[1], b[1], t)
    ]
  },

  /**
   * Interpolate between vectors a and b for a factor of t
   * with a smooth, ease in, ease out type effect
   * @param {number} a the start value
   * @param {number} b the end value
   * @param {number} t the time parameter
   * @returns {number} the interpolated value
   */
  smoothInterpolate(a, b, t) {
    const lerp = scalar.smoothInterpolate
    return [
      lerp(a[0], b[0], t),
      lerp(a[1], b[1], t)
    ]
  },

  /**
   * Interpolate with keyframes, e.g.,
   * [
   *   {stop: 0.1, value: [0, 100]},
   *   {stop: 0.5, value: [100, 100]},
   *   {stop: 0.9, value: [100, 0]}
   * ]
   * @param {keyframe[]} keyframes a list of keyframes
   * @param {number} time the time nearest the keyframe stop
   * @param {'linear'|'smooth'} timing interpolation timing
   * @param {object} lib the interpolation library to use
   */
  keyframeInterpolate(keyframes, time, timing='linear', lib=vector2d) {
    return keyframeInterpolate(keyframes, time, timing, lib)
  }
}

// -------------------------------------------------
// vector3d library
// -------------------------------------------------

const vector3d = {

  /**
   * Return the zero vector
   */
  zero() {
    return [0, 0, 0]
  },

  /**
   * Interpolate between vectors a and b for a factor of t
   * @param {number} a the start value
   * @param {number} b the end value
   * @param {number} t the time parameter
   * @returns {number} the interpolated value
   */
  linearInterpolate(a, b, t) {
    const lerp = scalar.linearInterpolate
    return [
      lerp(a[0], b[0], t),
      lerp(a[1], b[1], t),
      lerp(a[2], b[2], t)
    ]
  },

  /**
   * Interpolate between vectors a and b for a factor of t
   * with a smooth, ease in, ease out type effect
   * @param {number} a the start value
   * @param {number} b the end value
   * @param {number} t the time parameter
   * @returns {number} the interpolated value
   */
  smoothInterpolate(a, b, t) {
    const lerp = scalar.smoothInterpolate
    return [
      lerp(a[0], b[0], t),
      lerp(a[1], b[1], t),
      lerp(a[2], b[2], t)
    ]
  },

  /**
   * Interpolate with keyframes, e.g.,
   * [
   *   {stop: 0.1, value: [0, 100]},
   *   {stop: 0.5, value: [100, 100]},
   *   {stop: 0.9, value: [100, 0]}
   * ]
   * @param {keyframe[]} keyframes a list of keyframes
   * @param {number} time the time nearest the keyframe stop
   * @param {'linear'|'smooth'} timing interpolation timing
   * @param {object} lib the interpolation library to use
   */
  keyframeInterpolate(keyframes, time, timing='linear', lib=vector3d) {
    return keyframeInterpolate(keyframes, time, timing, lib)
  }
}


exports.util = util
exports.search = search
exports.scalar = scalar
exports.vector2d = vector2d
exports.vector3d = vector3d


src/example.js
const vk = require('./index')
const scalar = vk.scalar
const vector2d = vk.vector2d
const vector3d = vk.vector3d
const util = vk.util

function exScalar() {
  const keyframes = [{
    stop: 0,
    value: 0
  }, {
    stop: 0.5,
    value: 1
  }, {
    stop: 1,
    value: 0
  }]

  console.log('linear   (t=0):',
    scalar.keyframeInterpolate(keyframes, 0))

  console.log('linear (t=0.5):',
    scalar.keyframeInterpolate(keyframes, 0.5))

  console.log('linear   (t=1):',
    scalar.keyframeInterpolate(keyframes, 1))

  console.log('linear (t=0.1):',
    scalar.keyframeInterpolate(keyframes, 0.1))

  console.log('smooth (t=0.1):',
    scalar.keyframeInterpolate(keyframes, 0.1, 'smooth'))
}

function exVector2d() {
  const keyframes = [{
    stop: 0,
    value: [0, 100]
  }, {
    stop: 0.5,
    value: [100, 100]
  }, {
    stop: 1,
    value: [100, 0]
  }]

  console.log('linear   (t=0):',
    vector2d.keyframeInterpolate(keyframes, 0))

  console.log('linear (t=0.5):',
    vector2d.keyframeInterpolate(keyframes, 0.5))

  console.log('linear   (t=1):',
    vector2d.keyframeInterpolate(keyframes, 1))

  console.log('linear (t=0.1):',
    vector2d.keyframeInterpolate(keyframes, 0.1))

  console.log('smooth (t=0.1):',
    vector2d.keyframeInterpolate(keyframes, 0.1, 'smooth'))
}

function exVector3d() {
  const keyframes = [{
    stop: 0,
    value: [0, 0, 255]
  }, {
    stop: 0.5,
    value: [0, 255, 0]
  }, {
    stop: 1,
    value: [0, 0, 255]
  }]

  let result = vector3d.keyframeInterpolate(keyframes, 0.25)
  console.log('\nvector    (t=0.25):', result)
  console.log('rgb color (t=0.25):', util.vector3dToRgb(result))
  console.log('hex color (t=0.25):', util.vector3dToHex(result))

  result = vector3d.keyframeInterpolate(keyframes, 0.5)
  console.log('\vvector    (t=0.5):', result)
  console.log('rgb color (t=0.5):', util.vector3dToRgb(result))
  console.log('hex color (t=0.5):', util.vector3dToHex(result))

  result = vector3d.keyframeInterpolate(keyframes, 0.9)
  console.log('\nvector    (t=0.9):', result)
  console.log('rgb color (t=0.9):', util.vector3dToRgb(result))
  console.log('hex color (t=0.9):', util.vector3dToHex(result))
}

function exCustom() {
  const alpha = {

    zero() {
      return '*'
    },

    linearInterpolate(a, b, t) {
      return vk.scalar.linearInterpolate(a.charCodeAt(0), b.charCodeAt(0), t)
    },

    smoothInterpolate(a, b, t) {
      return vk.scalar.smoothInterpolate(a.charCodeAt(0), b.charCodeAt(0), t)
    },

    keyframeInterpolate(keyframes, t, timing='linear', lib=alpha) {
      const result = vk.scalar.keyframeInterpolate(keyframes, t, timing, lib)
      return String.fromCharCode(Math.round(result))
    }
  }

  const keyframes = [{
    stop: 0,
    value: 'a'
  }, {
    stop: 0.5,
    value: 'z'
  }, {
    stop: 1,
    value: 'a'
  }]

  console.log('linear   (t=0):',
    alpha.keyframeInterpolate(keyframes, 0, 'linear'))

  console.log('linear (t=0.5):',
    alpha.keyframeInterpolate(keyframes, 0.5, 'linear'))

  console.log('linear   (t=1):',
    alpha.keyframeInterpolate(keyframes, 1, 'linear'))

  console.log('linear (t=0.1):',
    alpha.keyframeInterpolate(keyframes, 0.1, 'linear'))

  console.log('smooth (t=0.1):',
    alpha.keyframeInterpolate(keyframes, 0.1, 'smooth'))
}

console.log('\nscalar')
console.log('--------')
exScalar()

console.log('\nvector2d')
console.log('--------')
exVector2d()

console.log('\nvector3d')
console.log('--------')
exVector3d()

console.log('\ncustom')
console.log('------')
exCustom()

console.log()
